File: backend/app/schemas/transit_fare.py
import uuid
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class TransitFareContributionCreate(BaseModel):
    origin: str = Field(..., min_length=1, max_length=255)
    destination: str = Field(..., min_length=1, max_length=255)
    mode: str = Field(..., pattern="^(cng|bus|train)$")
    fare_bdt: float = Field(..., ge=0)
    min_fare_bdt: Optional[float] = Field(default=None, ge=0)
    max_fare_bdt: Optional[float] = Field(default=None, ge=0)
    notes: Optional[str] = Field(default=None, max_length=2000)
    source_type: str = Field(default="observed", pattern="^(observed|quoted|booked)$")
    travel_date: Optional[date] = None


class TransitFareContributionRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    origin: str
    destination: str
    mode: str
    fare_bdt: float
    min_fare_bdt: Optional[float] = None
    max_fare_bdt: Optional[float] = None
    notes: Optional[str] = None
    source_type: str
    travel_date: Optional[date] = None
    submitted_at: datetime

    author_name: str
    author_picture_url: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class TransitFareContributionList(BaseModel):
    items: List[TransitFareContributionRead]
    total: int
    page: int
    per_page: int
    total_pages: int


class TransitFareModeEstimate(BaseModel):
    mode: str
    median_fare_bdt: Optional[float] = None
    submission_count: int
    recent_submission_count: int
    min_fare_bdt: Optional[float] = None
    max_fare_bdt: Optional[float] = None
    last_updated_at: Optional[datetime] = None
    sample_window_days: Optional[int] = None
    is_low_data: bool
    used_all_time_fallback: bool


class TransitFareEstimateResponse(BaseModel):
    origin: str
    destination: str
    estimates: List[TransitFareModeEstimate]


class BookingLink(BaseModel):
    id: str
    label: str
    url: str


class BookingLinksResponse(BaseModel):
    items: List[BookingLink]

File: frontend/src/types/transit-fare.ts
export type TransitFareMode = "cng" | "bus" | "train";
export type FareSourceType = "observed" | "quoted" | "booked";

export interface TransitFareModeEstimate {
  mode: TransitFareMode;
  median_fare_bdt: number | null;
  submission_count: number;
  recent_submission_count: number;
  min_fare_bdt: number | null;
  max_fare_bdt: number | null;
  last_updated_at: string | null;
  sample_window_days: number | null;
  is_low_data: boolean;
  used_all_time_fallback: boolean;
}

export interface TransitFareEstimate {
  origin: string;
  destination: string;
  estimates: TransitFareModeEstimate[];
}

export interface TransitFareContribution {
  id: string;
  user_id: string;
  origin: string;
  destination: string;
  mode: TransitFareMode;
  fare_bdt: number;
  min_fare_bdt: number | null;
  max_fare_bdt: number | null;
  notes: string | null;
  source_type: FareSourceType;
  travel_date: string | null;
  submitted_at: string;
  author_name: string;
  author_picture_url: string | null;
}

export interface TransitFareContributionList {
  items: TransitFareContribution[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface TransitFareContributionListParams {
  page?: number;
  per_page?: number;
  origin?: string;
  destination?: string;
  mode?: TransitFareMode;
  date_from?: string;
  date_to?: string;
}

export interface CreateTransitFareContributionPayload {
  origin: string;
  destination: string;
  mode: TransitFareMode;
  fare_bdt: number;
  min_fare_bdt?: number | null;
  max_fare_bdt?: number | null;
  notes?: string | null;
  source_type?: FareSourceType;
  travel_date?: string | null;
}

export interface BookingLink {
  id: string;
  label: string;
  url: string;
}

export interface BookingLinksResponse {
  items: BookingLink[];
}
